### PR TITLE
[AIRFLOW-6535] add exception to fail without retry

### PR DESCRIPTION
**Pull request by jstern**
I would like to be able to configure certain tasks to retry, but to also have a way to bypass retry if I can detect a condition that is unlikely to be change for the better.

For example, imagine I have a DAG with a large number of tasks that fetch data from an API ... sometimes the API returns a 200 and everything is fine, sometimes the API returns a 500 and I know this means that API is failing under load and I want to retry later, and sometimes the API returns a 400 indicating that I've configured my requests in a way that will never succeed. If the API returns a 400 for some/all of my tasks, then I have to wait for all of them to get through all their retries before the run fails.

My proposal is to add another subclass of AirflowException called AirflowFailException, and to update the exception handling in TaskInstance.run_raw_task such that when this exception is seen, the resulting behavior is that same as if we entered handle_failure and is_eligible_for_returned false (except with adjusted logging so it does not look like we simply hit our retry limit). 

---
Issue link: [AIRFLOW-6535](https://issues.apache.org/jira/browse/AIRFLOW-6535)

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: airflow/models/taskinstance.py [hunk 1]
```diff
@@ -1120,10 +1125,33 @@ def handle_failure(self, error, test_mode=None, context=None, session=None):
             # only mark task instance as FAILED if the next task instance
             # try_number exceeds the max_tries.
             if self.is_eligible_to_retry():
-                self.state = State.UP_FOR_RETRY
-                self.log.info('Marking task as UP_FOR_RETRY')
-                if task.email_on_retry and task.email:
-                    self.email_alert(error)
+                if force_fail is not True:
```
Comment (potiuk) [inline airflow/models/taskinstance.py]:
```suggestion
                if not force_fail:
```

or even better reverse the if and use `if force_fail:`

**Review by potiuk — COMMENTED** [review]
I like it, just one small comment. It looks like a good way. However it will also need some description in the documentation and some tests.

**Comment by jstern** [comment]
> I like it, just one small comment. It looks like a good way. However it will also need some description in the documentation and some tests.

Thanks for the quick review! Definitely will follow up with some tests shortly.

I was looking for a good place in the docs ... ideally I would have found an existing spot that discusses the semantics of AirflowSkipException and added something there, but the docs don't cover that either :/ Is there a particular spot within the documentation that would be best?

**Comment by jstern** [comment]
One other question: am I perhaps overcomplicating this? It occurs to me that if a task encounters an AirflowSkipException, we don't bother checking on its retry status or adjusting our messaging based on that:

```python
        except AirflowSkipException as e:
            # Recording SKIP
            # log only if exception has any arguments to prevent log flooding
            if e.args:
                self.log.info(e)
            self.refresh_from_db(lock_for_update=True)
            self.state = State.SKIPPED
            self.log.info(
                'Marking task as SKIPPED.'
                'dag_id=%s, task_id=%s, execution_date=%s, start_date=%s, end_date=%s',
                self.dag_id,
                self.task_id,
                self.execution_date.strftime('%Y%m%dT%H%M%S') if hasattr(
                    self,
                    'execution_date') and self.execution_date else '',
                self.start_date.strftime('%Y%m%dT%H%M%S') if hasattr(
                    self,
                    'start_date') and self.start_date else '',
                self.end_date.strftime('%Y%m%dT%H%M%S') if hasattr(
                    self,
                    'end_date') and self.end_date else '')
```

Maybe instead of messing with handle_failure I should just emulate that but with different messaging? Might be cleaner and easier to follow...

EDIT: but then I would still need to make sure failure email/callbacks happen ... so I guess I can either leave this structured the way I have it or include those in the except block.

**Comment by potiuk** [comment]
> > I like it, just one small comment. It looks like a good way. However it will also need some description in the documentation and some tests.
> 
> Thanks for the quick review! Definitely will follow up with some tests shortly.
> 
> I was looking for a good place in the docs ... ideally I would have found an existing spot that discusses the semantics of AirflowSkipException and added something there, but the docs don't cover that either :/ Is there a particular spot within the documentation that would be best?

I think "Concepts" is best and adding chapter about Exceptions would be good.

**Comment by potiuk** [comment]
> Maybe instead of messing with handle_failure I should just emulate that but with different messaging? Might be cleaner and easier to follow...

It's ok. I think it's easy enough.

**Comment by robinedwards** [comment]
Interesting I have a hack in one of my operators which allows the selection of specific exceptions to retry for.

**Comment by jstern** [comment]
> Interesting I have a hack in one of my operators which allows the selection of specific exceptions to retry for.

I'd be curious to see how you do it ... maybe it's not really that hacky? :) I'd be happy to close this PR if there's already a sensible/easy way to accomplish this within what airflow already provides...

File: airflow/models/taskinstance.py [hunk 2]
```diff
@@ -1120,10 +1125,33 @@ def handle_failure(self, error, test_mode=None, context=None, session=None):
             # only mark task instance as FAILED if the next task instance
             # try_number exceeds the max_tries.
             if self.is_eligible_to_retry():
-                self.state = State.UP_FOR_RETRY
-                self.log.info('Marking task as UP_FOR_RETRY')
-                if task.email_on_retry and task.email:
-                    self.email_alert(error)
+                if not force_fail:
```
Comment (houqp) [inline airflow/models/taskinstance.py]: perhaps we can make this if else combo less nested to reduce the complexity:

```python
if not force_fail and self.is_eligible_to_retry():
    self.state = State.UP_FOR_RETRY
    ...
else:
    if force_fail:
       info_msg = "Immediate failure requested; skipping retries and marking task as FAILED."
    elif task.retries:
       info_msg = "All retries failed; marking task as FAILED."
    else:
       info_msg = "Marking task as FAILED."
    self.log.info("%s.\ndag_id=%s, task_id=%s, execution_date=%s, start_date=%s, end_date=%s", info_msg, ...)
    if task.email_on_failure and task.email:
        self.email_alert(error)
```

I think it's reasonable to not check `task.retries` when force_fail is set and always show `Immediate failure requested; skipping retries and marking task as FAILED.`. If you still want to keep it, we can add that in `if force_fail` block.

Comment (jstern) [reply airflow/models/taskinstance.py]: yeah, the messiness of the nesting was weighing on me a bit when i made [this comment](https://github.com/apache/airflow/pull/7133#issuecomment-573155093) above, but what I was thinking of then wouldn't have worked because we still want all the callbacks to fire. I'll take a look at implementing your suggestion

**Comment by robinedwards** [comment]
@jstern https://gist.github.com/robinedwards/3f2ec4336e1ced084547d24d7e7ead3a

File: airflow/models/taskinstance.py [hunk 3]
```diff
@@ -1124,64 +1129,56 @@ def handle_failure(self, error, test_mode=None, context=None, session=None):
         if context is not None:
             context['exception'] = error
 
-        # Let's go deeper
-        try:
```
Comment (ashb) [inline airflow/models/taskinstance.py:1202]: What happened to this try/except?

File: airflow/models/taskinstance.py [hunk 4]
```diff
@@ -1103,7 +1108,7 @@ def _handle_reschedule(self, actual_start_date, reschedule_exception, test_mode=
         self.log.info('Rescheduling task, marking task as UP_FOR_RESCHEDULE')
 
     @provide_session
-    def handle_failure(self, error, test_mode=None, context=None, session=None):
+    def handle_failure(self, error, test_mode=None, context=None, session=None, force_fail=False):
```
Comment (ashb) [inline airflow/models/taskinstance.py]: A style point: please keep session as the final argument.

File: docs/conf.py [hunk 5]
```diff
@@ -207,9 +207,12 @@
 ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir))
 
 # Generate top-level
+
+allowed_top_level = ("exceptions.py",)
```
Comment (ashb) [inline docs/conf.py:216]: What is this all about?

File: tests/models/test_taskinstance.py [hunk 6]
```diff
@@ -1475,6 +1476,8 @@ def test_handle_failure(self):
                               dag=dag)
         ti1 = TI(task=task1, execution_date=start_date)
         ti1.state = State.FAILED
+        ti1.start_date = start_date
+        ti1.end_date = end_date
```
Comment (ashb) [inline tests/models/test_taskinstance.py]: Why was this change needed? The tests should pass without this change, no?

**Review by ashb — COMMENTED** [review]
Nice idea.

The one thing I don't think the tests cover is that actually raising this new exception causes the task to not be retried -- you've tested the change to handle_failure, but that is essentially the internal details, rather than the "interface" of the new feature.

File: airflow/models/taskinstance.py (hunk 3 above)
Comment (jstern) [reply airflow/models/taskinstance.py:1202]: the try/excepts have been tightened in scope to only be around the conditions they are logging and moving on from. see my response to your question about the start/end dates in the tests for more color on that

File: airflow/models/taskinstance.py (hunk 4 above)
Comment (jstern) [reply airflow/models/taskinstance.py]: 10-4, will change!

File: docs/conf.py (hunk 5 above)
Comment (jstern) [reply docs/conf.py:216]: ah, was wondering if this would come up. @potiuk suggested that I should add documentation to the concepts section, and it felt incomplete to me to add that without linking to the exceptions api docs. all modules at the top level (airflow/*.py) are excluded from the documentation and my link to the exceptions api doc could not be rendered and was breaking the docs build. adding this ... exception :) ... fixed that

i'm happy to take any direction whether that's leaving this as-is, removing the links to the exception api docs, or something in-between. just let me know what you'd prefer

File: tests/models/test_taskinstance.py (hunk 6 above)
Comment (jstern) [reply tests/models/test_taskinstance.py]: the original structure of the method under test had a wide scope for the try/except block that was intended to log when we're unable to send an email and move on. i think TypeError's due to these attributes being none were being caught by that block and allowing the test to complete without error

**Comment by jstern** [comment]
> The one thing I don't think the tests cover is that actually raising this new exception causes the task to not be retried -- you've tested the change to handle_failure, but that is essentially the internal details, rather than the "interface" of the new feature.

I'll take some time to look deeper but I didn't find any existing tests of the interface for handling the other exceptions, so I glommed onto the closest test I could find. If you know offhand where better example tests that I can extend please point me to them :) Or I can try to take some time to add them.

File: docs/conf.py (hunk 5 above)
Comment (jstern) [reply docs/conf.py:216]: i can also just add a comment explaining why this exists

File: tests/models/test_taskinstance.py (hunk 6 above)
Comment (jstern) [reply tests/models/test_taskinstance.py]: for more context, here is the failure that cropped up during the execution of existing test code after I narrowed the scope of the try/except to only be around sending the email. my mistake, it was an AttributeError not a TypeError :)

```
        self.log.info(
            '%s dag_id=%s, task_id=%s, execution_date=%s, start_date=%s, end_date=%s',
            log_message,
            self.dag_id,
            self.task_id,
            self.execution_date.strftime('%Y%m%dT%H%M%S'),
>           self.start_date.strftime('%Y%m%dT%H%M%S'),
            self.end_date.strftime('%Y%m%dT%H%M%S')
        )
E       AttributeError: 'NoneType' object has no attribute 'strftime'
```

Comment (jstern) [reply tests/models/test_taskinstance.py]: Aha! https://github.com/apache/airflow/pull/7257

File: airflow/models/taskinstance.py (hunk 3 above)
Comment (jstern) [reply airflow/models/taskinstance.py:1202]: To give a more complete answer, here is the try/except in master. I've elided some arguments to log.info and added comments to the two lines where an exception being raised **would** result in an accurately logged error message. Exceptions raised on other lines can cause us to log a message about failing to send an email when that's not what actually happened.

```python
        # Let's go deeper
        try:
            # Since this function is called only when the TaskInstance state is running,
            # try_number contains the current try_number (not the next). We
            # only mark task instance as FAILED if the next task instance
            # try_number exceeds the max_tries.
            if self.is_eligible_to_retry():
                self.state = State.UP_FOR_RETRY
                self.log.info('Marking task as UP_FOR_RETRY')
                if task.email_on_retry and task.email:
                    self.email_alert(error)  # RELEVANT TO EXCEPT BLOCK
            else:
                self.state = State.FAILED
                if task.retries:
                    self.log.info(...)
                else:
                    self.log.info(...)
                if task.email_on_failure and task.email:
                    self.email_alert(error)  # RELEVANT TO EXCEPT BLOCK
        except Exception as e2:
            self.log.error('Failed to send email to: %s', task.email)
            self.log.exception(e2)
```

And here is the more tightly-scoped version in this PR:

```python
    if email_for_state and task.email:
            try:
                self.email_alert(error)
            except Exception as e2:
                self.log.error('Failed to send email to: %s', task.email)
                self.log.exception(e2)
```

While https://github.com/apache/airflow/pull/7257 addresses the only non-email failure I've personally seen happen in practice, this PR does put some lines of code that were previously inside of a try/except on the outside (and I'm sure what I've personally seen happen is a fraction of what actually might). Please let me know if this increased chance of other exceptions bubbling up uncaught is unacceptable and I'll look to restructure while trying to preserve accurate logging....

Comment (ashb) [reply airflow/models/taskinstance.py:1202]: Does is_eligable_for_retry do anything more than check instance attributes? If not then this change is probably okay

File: docs/conf.py (hunk 5 above)
Comment (ashb) [reply docs/conf.py:216]: Ah, the API is handled differently (we render it with an autodoc extension) - can you get away with just linking to

```
   :py.module:`airflow.exceptions`
```

where you want to and then this change might not be needed?

Comment (ashb) [reply docs/conf.py:216]: Ah, the API is handled differently (we render it with an autodoc extension) - can you get away with just linking to

```
   :py.module:`airflow.exceptions`
```

where you want to and then this change might not be needed?

File: airflow/models/taskinstance.py (hunk 3 above)
Comment (jstern) [reply airflow/models/taskinstance.py:1202]: Here's the entirety of it:

```python
    def is_eligible_to_retry(self):
        """Is task instance is eligible for retry"""
        return self.task.retries and self.try_number <= self.max_tries
```

File: docs/conf.py (hunk 5 above)
Comment (jstern) [reply docs/conf.py:216]: This works ... 

```
:py:mod:`airflow.exceptions`
```

... but only if I keep my changes in conf.py 😦 Otherwise it "works" but does not create a link. I'm pretty sure the code I'm touching in conf.py is actively excluding that module from the autodocs.

File: airflow/models/taskinstance.py (hunk 3 above)
Comment (ashb) [reply airflow/models/taskinstance.py:1202]: Thanks (I was reviewing from my phone, so wasn't easy for me to check)

File: docs/conf.py (hunk 5 above)
Comment (ashb) [reply docs/conf.py:216]: @mik-laj @potiuk Any thoughts on this specific change?

Comment (ashb) [reply docs/conf.py:216]: Looking at it a bit more I think you are right though @jstern

File: tests/models/test_taskinstance.py (hunk 6 above)
Comment (ashb) [reply tests/models/test_taskinstance.py]: With that PR is this change now not needed? (Does it make sense to keep it anyway?)

Comment (jstern) [reply tests/models/test_taskinstance.py]: Actually, I needed to do something to make sure my boiling down of the logging.info call into one spot also incorporated that safety check. I've done that now and removed the start_date/end_date assignments from the test again

File: docs/conf.py (hunk 5 above)
Comment (jstern) [reply docs/conf.py:216]: @ashb @potiuk @mik-laj - any thoughts about whether it's ok to keep this change to include the exceptions module in the docs?

---

On a related note, after updating this pr from master, I'm getting a new docs build failure:

```
/opt/airflow/docs/_api/airflow/providers/amazon/aws/hooks/datasync/index.rst:66: WARNING: more than one target found for cross-reference 'AirflowException': airflow.exceptions.AirflowException, airflow.providers.yandex.hooks.AirflowException
/opt/airflow/docs/_api/airflow/providers/amazon/aws/operators/datasync/index.rst:167: WARNING: more than one target found for cross-reference 'AirflowException': airflow.exceptions.AirflowException, airflow.providers.yandex.hooks.AirflowException
/opt/airflow/docs/_api/airflow/providers/amazon/aws/operators/datasync/index.rst:167: WARNING: more than one target found for cross-reference 'AirflowException': airflow.exceptions.AirflowException, airflow.providers.yandex.hooks.AirflowException
/opt/airflow/docs/_api/airflow/providers/amazon/aws/operators/datasync/index.rst:167: WARNING: more than one target found for cross-reference 'AirflowException': airflow.exceptions.AirflowException, airflow.providers.yandex.hooks.AirflowException
/opt/airflow/docs/_api/airflow/providers/amazon/aws/operators/datasync/index.rst:167: WARNING: more than one target found for cross-reference 'AirflowException': airflow.exceptions.AirflowException, airflow.providers.yandex.hooks.AirflowException
```

I can make this go away by removing my change that allows `airflow.exceptions` api docs to be built, but this does not seem like a satisfactory course of action, because the `airflow.providers.yandex.hooks.AirflowException` will remain in the docs and it shouldn't. It's there because somehow these lines in `airflow/providers/yandex/hooks/__init__.py`:

```
from airflow.exceptions import AirflowException
from airflow.hooks.base_hook import BaseHook
```

are causing autoapi to include `AirflowException` and `BaseHook` in the docs for `airflow.providers.yandex.hooks` as if those classes were defined in that module.

Screen Shot 2020-02-27 at 4 25 31 PM

In other words, autoapi is generating this in `docs/_api/airflow/providers/yandex/hooks/index.rst`:

```
Package Contents
----------------

.. py:exception:: AirflowException

   Bases: :class:`Exception`

   Base class for all Airflow's errors.
   Each custom exception should be derived from this class

   .. attribute:: status_code
      :annotation: = 500

.. py:class:: BaseHook
```

I don't know if this is a bug in autoapi or in how the airflow project configures it, and I don't know why the warnings are only about AirflowException and not BaseHook...

**Comment by KevinYang21** [comment]
Hi @jstern , this is a very nice feature that we are also very interested in. Do we still plan to merge it? Anything we can do to help?

**Comment by jstern** [comment]
Thanks @KevinYang21 - I definitely would like to see it get merged, but I'm stuck on a strange problem with the docs build throwing warnings due to the problem I described in https://github.com/apache/airflow/pull/7133#discussion_r389372194 above. 

The short version of the issue is that my PR adds `airflow.exceptions` to the API docs, but this causes warnings in the doc build because `AirflowException` is already included (incorrectly) in the yandex provider API docs. I have not yet had the time or inspiration to figure out why that bug exists in master or how to correct it.

Another option that would let the build succeed would be to revert the change that adds the exceptions module to the api docs, but since I believe the docs in master are wrong I have been reluctant to take this step.

Perhaps I should file a JIRA ticket for the docs bug in master and hope someone who understands Sphinx and autoapi better than I do knows how to fix it?

**Comment by KevinYang21** [comment]
@jstern sry I was distracted, yes I think your plan is good. Saw that you got a PR fixing doc, will look into that one too. Thank you!

**Comment by tooptoop4** [comment]
gentle ping @jstern

**Comment by jstern** [comment]
> gentle ping @jstern

hi @tooptoop4 ! just waiting on https://github.com/apache/airflow/pull/8095 to hit master so the docs can build cleanly and then hopefully we'll be able to get this one ready to merge

**Comment by tooptoop4** [comment]
bump?

**Review by tooptoop4 — APPROVED** [review]

**Comment by jstern** [comment]
@KevinYang21 if you have some time to look at this again ... nothing I'm aware of to hold it up other than committer approval.

cc @ashb @potiuk ... i know you guys are very busy but i'd be very grateful if you could find time to take another look at this :)

File: docs/concepts.rst [hunk 7]
```diff
@@ -1216,6 +1216,49 @@ template string:
...
+behavior is desired:
+
+* ``AirflowSkipException`` can be raised to set the state of the current task instance to "skipped"
+* ``AirflowFailException`` can be raised to set the state of the current task to "failed" regardless
+  of whether there are any retry attempts remaining.
+
+This example illustrates some possibilities
+
+.. code:: python
+
```
Comment (ashb) [inline docs/concepts.rst:1238]: can you add the import here? `from airflow.exceptions import ...`

**Review by ashb — CHANGES_REQUESTED** [review]
one tiny doc change then I'm happy with this.

sorry you had to chase so many times!

File: docs/concepts.rst (hunk 7 above)
Comment (jstern) [reply docs/concepts.rst:1238]: done :)

**Review by ashb — APPROVED** [review]

**Comment by KevinYang21** [comment]
🎉 I need to really spend some time on my email filter, keep missing such important PR @ :) Thank you @jstern

**Comment by marshall7m** [comment]
Hi there, just wondering what version of airflow the AirflowFailException is available on. Thanks!

**Comment by ashb** [comment]
It's marked to be included in 1.10.11

image
